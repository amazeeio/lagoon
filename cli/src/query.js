// @flow

import path from 'path';
import os from 'os';
import url from 'url';
import R from 'ramda';
import { config } from './config/';
import { fileExists, readFile } from './util/fs';
import request from './util/request';
import { printErrors } from './printErrors';

type QLQueryArgs = {
  cerr: typeof console.error,
  query: string,
  variables?: Object,
  headers?: Object,
  pretty?: boolean,
};

export async function runGQLQuery({
  cerr,
  query,
  variables,
  headers: customHeaders = {},
  pretty = false,
}:
QLQueryArgs): Object {
  const headers = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    ...customHeaders,
  };

  if (!headers.Authorization) {
    const tokenFile = path.join(os.homedir(), '.lagoon-token');
    const tokenFileExists = await fileExists(tokenFile);

    if (tokenFileExists) {
      const tokenBuffer = await readFile(tokenFile);
      const token = tokenBuffer.toString().replace(/(\r\n|\n|\r)/gm, '');
      headers.Authorization = `Bearer ${token}`;
    }
  }

  const apiUrl = R.prop('api', config) || 'https://api.amazee.io/graphql';

  const {
    hostname, path: pathname, port: urlPort, protocol,
  } = url.parse(
    apiUrl,
  );

  if (hostname == null) {
    throw new Error(
      'API URL configured under the "api" key in .lagoon.yml doesn\'t contain a valid hostname.',
    );
  }

  const body = JSON.stringify(
    {
      query,
      variables,
    },
    null,
    pretty ? 2 : 0,
  );

  const protocolPorts = {
    'https:': 443,
    'http:': 80,
  };

  const options = {
    hostname,
    path: pathname,
    port:
      urlPort === null
        ? R.propOr(443, protocol)(protocolPorts)
        : Number(urlPort),
    method: 'POST',
    headers,
    body,
    rejectUnauthorized: false,
  };

  try {
    return await request(options);
  } catch (err) {
    const error = R.ifElse(
      // For socket hang ups...
      R.propEq('message', 'socket hang up'),
      // ...print a nicer error message...
      R.always('Could not connect to API.'),
      // ...otherwise just return the error message
      R.prop('message'),
    )(err);
    printErrors(cerr, error);
    process.exit(1);
  }
}

export default runGQLQuery;
