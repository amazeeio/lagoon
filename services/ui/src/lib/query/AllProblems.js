import gql from 'graphql-tag';

export default gql`
  {
    allGroups {
      name
      projects {
        name
        id
        environments(type: PRODUCTION) {
          name
          id
          problems {
            id
            identifier
            severity
            source
          }
        }
      }
    }
  }
`;
