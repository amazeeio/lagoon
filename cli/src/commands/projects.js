// @flow

import { table } from 'table';
import R from 'ramda';

import gql from '../gql';
import { runGQLQuery } from '../query';
import { printGraphQLErrors } from '../printErrors';

import typeof Yargs from 'yargs';
import type { BaseArgs } from '.';

export const command = 'projects';
export const description = 'List all projects';

export function builder(yargs: Yargs) {
  return yargs
    .usage(`$0 ${command} - ${description}`)
    .example(`$0 ${command}`, 'List all projects');
}

type ListProjectsArgs = {
  clog: typeof console.log,
  cerr: typeof console.error,
};

export async function listProjects({
  clog,
  cerr,
}:
ListProjectsArgs): Promise<number> {
  const query = gql`
    query AllProjects {
      allProjects {
        name
        git_url
        branches
        pullrequests
        created
      }
    }
  `;

  const result = await runGQLQuery({
    cerr,
    query,
  });

  const { errors } = result;
  if (errors != null) {
    return printGraphQLErrors(cerr, ...errors);
  }

  const sortByName = R.sortBy(R.compose(R.toLower, R.propOr('', 'name')));

  const projects = R.compose(sortByName, R.pathOr([], ['data', 'allProjects']))(
    result,
  );

  if (projects.length === 0) {
    clog('No projects found.');
    return 0;
  }

  clog(
    table([
      ['Project', 'Git URL', 'Branches', 'Pull Requests', 'Created'],
      ...R.map(
        project => [
          project.name,
          project.git_url,
          String(project.branches),
          String(project.pullrequests),
          project.created,
        ],
        projects,
      ),
    ]),
  );

  return 0;
}

export async function handler({ clog, cerr }: BaseArgs): Promise<number> {
  return listProjects({ clog, cerr });
}
