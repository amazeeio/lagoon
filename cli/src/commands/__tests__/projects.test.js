// @flow

import { runGQLQuery } from '../../query';
import { listProjects } from '../projects';

jest.mock('../../query');

const _mock = (mockFn: any): JestMockFn => mockFn;

describe('listProjects', () => {
  it('should list details for multiple projects', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() =>
      Promise.resolve({
        data: {
          allProjects: [
            {
              name: 'credentialstest-project1',
              git_url: 'ssh://git@192.168.99.1:2222/git/project1.git',
              branches: 'true',
              pullrequests: null,
              created: '2018-01-15 11:09:35',
            },
            {
              name: 'credentialstest-project2',
              git_url: 'ssh://git@192.168.99.1:2222/git/project2.git',
              branches: 'true',
              pullrequests: null,
              created: '2018-01-15 11:09:35',
            },
          ],
        },
      }),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await listProjects({
      clog,
      cerr,
    });

    expect(code).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });

  it('should print notice on empty projects array', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() => Promise.resolve([]));

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await listProjects({
      clog,
      cerr,
    });

    expect(code).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });

  it('should display error, if GraphQL sends error messages', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() =>
      Promise.resolve({
        errors: [{ message: 'something something error' }],
      }),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await listProjects({
      clog,
      cerr,
    });

    expect(code).toBe(1);
    expect(cerr.mock.calls).toMatchSnapshot();
  });
});
