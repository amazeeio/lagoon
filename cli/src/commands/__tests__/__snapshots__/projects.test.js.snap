// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`listProjects should display error, if GraphQL sends error messages 1`] = `
Array [
  Array [
    "[31mOops! The lagoon API returned an error:[39m",
  ],
  Array [
    "[31m{[39m
[31m  \\"message\\": \\"something something error\\"[39m
[31m}[39m",
  ],
]
`;

exports[`listProjects should list details for multiple projects 1`] = `
Array [
  Array [
    "╔══════════════════════════╤══════════════════════════════════════════════╤══════════╤═══════════════╤═════════════════════╗
║ Project                  │ Git URL                                      │ Branches │ Pull Requests │ Created             ║
╟──────────────────────────┼──────────────────────────────────────────────┼──────────┼───────────────┼─────────────────────╢
║ credentialstest-project1 │ ssh://git@192.168.99.1:2222/git/project1.git │ true     │ null          │ 2018-01-15 11:09:35 ║
╟──────────────────────────┼──────────────────────────────────────────────┼──────────┼───────────────┼─────────────────────╢
║ credentialstest-project2 │ ssh://git@192.168.99.1:2222/git/project2.git │ true     │ null          │ 2018-01-15 11:09:35 ║
╚══════════════════════════╧══════════════════════════════════════════════╧══════════╧═══════════════╧═════════════════════╝
",
  ],
]
`;

exports[`listProjects should print notice on empty projects array 1`] = `
Array [
  Array [
    "No projects found.",
  ],
]
`;
